app/src/lib/stores/app-store.ts

```diff
@@ -5294,119 +5295,74 @@ export class AppStore extends TypedBaseStore<IAppState> {
   public async _checkoutPullRequest(
     repository: RepositoryWithGitHubRepository,
     prNumber: number,
-    ownerLogin: string,
+    headRepoOwner: string,
     headCloneUrl: string,
     headRefName: string
   ): Promise<void> {
-    const branch = await this._getPullRequestHeadBranchInRepo(
-      repository,
-      headCloneUrl,
-      headRefName
-    )
-
-    // N.B: This looks weird, and it is. _checkoutBranch used
-    // to behave this way (silently ignoring checkout) when given
-    // a branch name string that does not correspond to a local branch
-    // in the git store. When rewriting _checkoutBranch
-    // to remove the support for string branch names the behavior
-    // was moved up to this method to not alter the current behavior.
-    //
-    // https://youtu.be/IjmtVKOAHPM
-    if (branch !== null) {
-      await this._checkoutBranch(repository, branch)
-      this.statsStore.recordPRBranchCheckout()
-      return
-    }
-
-    const remoteName = forkPullRequestRemoteName(ownerLogin)
+    const gitStore = this.gitStoreCache.get(repository)
     const remotes = await getRemotes(repository)
-    const remote =
-      remotes.find(r => r.name === remoteName) ||
-      (await addRemote(repository, remoteName, headCloneUrl))
+    const forkRemoteName = forkPullRequestRemoteName(headRepoOwner)
 
-    if (remote.url !== headCloneUrl) {
-      const error = new Error(
-        `Expected PR remote ${remoteName} url to be ${headCloneUrl} got ${remote.url}.`
-      )
-
-      log.error(error.message)
-      return this.emitError(error)
-    }
+    // Find an existing remote (regardless if set up by us or outside of
+    // Desktop). If we can't find one we'll create a Desktop for remote.
+    const remote =
+      remotes.find(r => urlMatchesRemote(headCloneUrl, r)) ??
+      (await addRemote(repository, forkRemoteName, headCloneUrl))
```

Comment:
This is something you figured out, so I won't take any credit for it 😂 , but I thought it was worth documenting it here maybe?

So as you explained to me, this `addRemote` could fail (silently as an "unhandled promise rejection" instead of crashing) if the repo already had a remote set up with this `forkRemoteName` name, but a different URL. For example, in the weird scenario where someone used Desktop (without the `ownerLogin` vs `headRepoOwner` changes in this PR) like this:

1. Let's say user `A` submits a PR to the repo `R`, using a branch that comes from the fork `FZ`, which belongs to user `Z`.
1. Then I use Desktop (prod, without this fixes) to checkout to that PR's branch, I'd get a remote called `github-desktop-A` pointing to the fork `FZ` (which belongs to user `Z`, not to user `A`).
1. Then I update Desktop to get the fixes from this PR.
1. After that I use Desktop again now to checkout to some other PR that `A` submitted to `R`, but now the source branch comes from the fork `FA` which actually belongs to `A`.
1. In that case, Desktop will look for a remote referencing `FA` using its clone URL, it will find nothing, so it will **try** to add, thanks to the fixes here, a remote like `github-desktop-A`, and it will fail because `github-desktop-A` was created previously to reference `FZ` instead 😢 

Of course, there is also the possibility of users tinkering manually with git remotes, which could end up in the same scenario but in a less confusing way 😂 

TLDR; should we catch and handle the error here as we discussed?

Reply:
:heart_eyes: This is so well documented. I don't believe my explanation was nearly as easy to follow. Thanks for taking the time to lay it out here

> TLDR; should we catch and handle the error here as we discussed?

Yeah I think it's worth it even though it's likely super rare. I'll get on it